### PR TITLE
Refactoring UploadRequestGenerator for Upload Sync Redesign

### LINKED ISSUE #2013: SDC : Answer being vanished for Quantity type item after packRepeatedGroups
**Describe the bug**
For nested item of group type with enable expression, in some case, answer is getting vanished for the item after packRepeatedGroups method while processing QR

**Component**
SDC library

**To Reproduce**
1. On Master branch apply the patch [signs_complete_patch.txt](https://github.com/google/android-fhir/files/11542871/signs_complete_patch.txt)
2. Run the app, Go to new questionnaire (+ icon n patient list)
3. Answer is not being rendered for the Respiratory rate
4. Undo the patch changes and apply this patch: [signs_questionnaire_patch_mvp.txt](https://github.com/google/android-fhir/files/11543243/signs_questionnaire_patch_mvp.txt)
 list)
5. Now Re-run the app and again open the questionnaire, Interestingly it's working fine with MVP signs_questionnaire_patch_mvp.txt for the same item

**Expected behavior**
Respiratory rate item answer should be rendered if it's given in QR, while debugging through SDC methods , we found that answer object for Respiratory Rate item is being vanished after packRepeatedGroups method with the Patch [signs_complete_patch.txt](https://github.com/google/android-fhir/files/11542871/signs_complete_patch.txt) 

**Screenshots**
Answer no being rendered for Respiratory rate
image

Answer being rendered for Respiratory rate as standalone item
image

**Smartphone (please complete the following information):**
 - Device: Pixel4 emulator

**Additional context**
After applying patch: Questionnaire is at file: assets/new-patient-registration-paginated.json & QuestionnaireResponse is on line: 83(AddPatientFragment.kt)

@jingtang10 @santosh-pingle

### PR DESCRIPTION
**Pull request by anchita-g**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

**Description**

- Refactored `Request` into `DownloadRequest` and `UploadRequest` and moved them into appropriate packages.
- Refactored `IndividualRequestGenerator` to `UrlRequestGenerator` to bring it more in line with the type of `UploadRequest`
- Created `UploadRequestGeneratorMode` to support creation of appropriate `UploadRequestGenerator` 

**Type**
Choose one: Feature

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

File: engine/src/main/java/com/google/android/fhir/sync/upload/request/UploadRequestGeneratorMode.kt [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ * limitations under the License.
+ */
+
+package com.google.android.fhir.sync.upload.request
+
+import org.hl7.fhir.r4.model.Bundle
+import org.hl7.fhir.r4.model.codesystems.HttpVerb
+
+/** Mode to decide the type of [UploadRequest] that needs to be generated */
+sealed class UploadRequestGeneratorMode {
```
Comment (omarismail94) [inline engine/src/main/java/com/google/android/fhir/sync/upload/request/UploadRequestGeneratorMode.kt:23]: does this need to be in a separate class, or can we move it to UploadRequestGenerator?

Comment (omarismail94) [inline engine/src/main/java/com/google/android/fhir/sync/upload/request/UploadRequestGeneratorMode.kt:23]: can you mark it as internal as we don't want to expose this?

Comment (anchita-g) [reply engine/src/main/java/com/google/android/fhir/sync/upload/request/UploadRequestGeneratorMode.kt:23]: Do we follow a particular style on when when to keep multiple classes in a single file. Do we not keep one class declaration per source file

Comment (omarismail94) [reply engine/src/main/java/com/google/android/fhir/sync/upload/request/UploadRequestGeneratorMode.kt]: Not really, but if it's used in one file only, then we keep it in the same file. See LocalChangeFetcher as an example

Comment (anchita-g) [reply engine/src/main/java/com/google/android/fhir/sync/upload/request/UploadRequestGeneratorMode.kt]: Sure, moved it.

**Review by jingtang10 — APPROVED** [review]

**Pull request merged**